src/datasources/accounts/accounts.datasource.spec.ts

```diff
@@ -149,4 +150,150 @@ describe('AccountsDatasource tests', () => {
+        expect.arrayContaining(
+          accountDataSettings.map((ads) => ({
+            account_id: account.id,
+            account_data_type_id: dataTypes.find(
+              (dt) => dt.name === ads.dataTypeName,
+            )?.id,
+            enabled: ads.enabled,
+            created_at: expect.any(Date),
+            updated_at: expect.any(Date),
+          })),
+        ),
+      );
+
+      const accountDataSettings2 = accountDataSettings.map((ads) => ({
+        ...ads,
+        enabled: !ads.enabled,
+      }));
+      const upsertAccountDataSettingsDto2 =
+        upsertAccountDataSettingsDtoBuilder()
+          .with('accountDataSettings', accountDataSettings2)
+          .build();
+
+      const afterUpdate = await target.upsertAccountDataSettings(
+        address,
+        upsertAccountDataSettingsDto2,
+      );
+
+      expect(afterUpdate).toStrictEqual(
+        expect.arrayContaining(
+          accountDataSettings.map((ads) => ({
+            account_id: account.id,
+            account_data_type_id: dataTypes.find(
+              (dt) => dt.name === ads.dataTypeName,
+            )?.id,
+            enabled: !ads.enabled, // 'enabled' row was updated
+            created_at: expect.any(Date),
+            updated_at: expect.any(Date),
+          })),
+        ),
+      );
+    });
+
+    it('throws an error if the account does not exist', async () => {
+      const address = getAddress(faker.finance.ethereumAddress());
+      const upsertAccountDataSettingsDto =
+        upsertAccountDataSettingsDtoBuilder().build();
+      const { accountDataSettings } = upsertAccountDataSettingsDto;
+      const dataTypeNames = accountDataSettings.map((ads) => ({
+        name: ads.dataTypeName,
+      }));
+      await sql`INSERT INTO account_data_types ${sql(dataTypeNames, 'name')} returning *`;
+
+      await expect(
+        target.upsertAccountDataSettings(address, upsertAccountDataSettingsDto),
+      ).rejects.toThrow('Error getting account.');
+    });
+
+    it('throws an error if a non-existent data type is provided', async () => {
+      const address = getAddress(faker.finance.ethereumAddress());
+      const upsertAccountDataSettingsDto =
+        upsertAccountDataSettingsDtoBuilder().build();
+      const { accountDataSettings } = upsertAccountDataSettingsDto;
+      await target.createAccount(address);
+      const dataTypeNames = accountDataSettings.map((ads) => ({
+        name: ads.dataTypeName,
+      }));
+      await sql`INSERT INTO account_data_types ${sql(dataTypeNames, 'name')} returning *`;
+      upsertAccountDataSettingsDto.accountDataSettings.push({
+        dataTypeName: faker.lorem.slug(),
+        enabled: faker.datatype.boolean(),
+      });
+
+      await expect(
+        target.upsertAccountDataSettings(address, upsertAccountDataSettingsDto),
+      ).rejects.toThrow('Invalid data type.');
+    });
+
+    it('throws an error if an inactive data type is provided', async () => {
```

Comment:
This extends upon my other comment regarding this flag. Do we want to prevent the client from "registering" for settings outright? We need to consider _how_ this is going to be toggled. It's somewhat of an admin-related flag.

Reply:
(Related to https://github.com/safe-global/safe-client-gateway/pull/1715#discussion_r1668434862)

In my opinion and considering the changes in the data types will be likely uncommon, I'd opt for doing this via database migrations. And I'd include a new FF in the service `FF_RUN_MIGRATIONS`, so it would give us some extra control regarding that.